### PR TITLE
Blog implementation

### PR DESCRIPTION
**Pull request by matt-hernandez**
Here is a PR for the new blog from a branch made from the original repo. This is the exact same code as #91, which I'm going to close. I squashed the commits when making this new branch from the previous one, so we're starting from a pretty clean history to make suggestions and add new code.

@colbywhite @brookslybrand If you want to make any contributions, do so on this branch now. There are already a few basic tests added. Any other suggestions are welcome.

### CODE REVIEW

**Comment by matt-hernandez** [comment]
@brookslybrand The Github action said it had a problem deploying, but I can navigate to https://remixaustin-com-pr-99.fly.dev/ and see the site. So I'm guessing the error, while still needing to be fixed, didn't stop us from seeing the site with the blog. [And it appears to be working!!](https://remixaustin-com-pr-99.fly.dev/blog)

File: blog/writeFrontMatterCache.test.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { validateFrontMatter } from "./writeFrontMatterCache";
+
+describe("writeFrontMatterCache", () => {
```
Comment (colbywhite) [inline blog/writeFrontMatterCache.test.ts:3]: this was the first set of tests i was working on writing the other day. my initial thought to seeing this version is that the fact that `validateFrontMatter` and `buildFrontMatterCache` are exported but _only_ used in tests is a code smell in my opinion. but that's more of a philosophical thought and i'd rather not bog you down with testing philosophy back-and-forth (especially since the bigger picture is that we understandably don't have a project-wide testing style here). 

so i'm going to hold off on writing tests around this and instead i now propose we focus on getting this merged asap. and for testing, i'm going to follow this up with a larger effort to add more testing to this project as a whole and analyzing the tests that are already there. 

thoughts on that strategy? @brookslybrand?

Comment (colbywhite) [reply blog/writeFrontMatterCache.test.ts:3]: maybe a better way to ask that question is what are the dealbreakers preventing this PR from getting merged? i normally would say testing, but i'm proposing hold off for a bit cuz i'm going to make sure we fast follow.

File: app/routes/resource/get-recent-front-matter.ts
Comment (colbywhite) [inline app/routes/resource/get-recent-front-matter.ts:1]: i still vote that this should be combined w/ `get-all-front-matter` and use paging params to get the recent stuff. 

dealbreaker? i'll cede to @brookslybrand

File: app/utils.ts [hunk 2]
```diff
@@ -71,6 +71,13 @@ export function isEmptyString(val: unknown): boolean {
   return typeof val === "string" && val.trim().length <= 0;
 }
 
+export const publishDateFormatter = new Intl.DateTimeFormat("en-US", {
```
Comment (colbywhite) [inline app/utils.ts:74]: typically constants are caps, and the other constants in this file follow that convention.

```suggestion
export const PUBLISH_DATE_FROMATTER = new Intl.DateTimeFormat("en-US", {
```

File: blog/buildPosts.ts
Comment (colbywhite) [inline blog/buildPosts.ts:1]: this is def _not_ a dealbreaker but j/ something i wanted to put on your mind for when we export this stuff to a lib. we can likely combine `blogDevWatch` and `buildPosts` into a single CLI. something along the lines of `remix-blog build` and `remix-blog build --watch`. this would help w/ some of the redundancy in these two files (and potentially make it easier to test).

File: blog/writeFrontMatterCache.test.ts (hunk 1 above)
Comment (matt-hernandez) [reply blog/writeFrontMatterCache.test.ts:3]: The reason for those exports is just to use them in testing, and pretty much no other reason. I did that for familiarity's sake because I've never in my career had to write tests that check the file system and assert on a file's expected output. It feels weird to me, but maybe that's just my front-end bias speaking. If you think that's a valid test to write, I would say you can adjust this as needed.

File: e2e/homepage.spec.ts [hunk 3]
```diff
@@ -22,3 +22,33 @@ test.describe("Homepage", () => {
     await expect(page.getByText(copyrightText, { exact: true })).toBeVisible();
   });
 });
+
+test.describe("Blog", () => {
```
Comment (colbywhite) [inline e2e/homepage.spec.ts:26]: let's move this out of `homepage.spec` since this isn't about the homepage.

Comment (matt-hernandez) [reply e2e/homepage.spec.ts:26]: I'll move this to its own file. And I'll hold off on making any more tests or adjustments to tests afterward. I do want to see your take on how these tests should be done, considering you may have experience working on more back-end systems than me in my time.

File: app/utils.ts (hunk 2 above)
Comment (matt-hernandez) [reply app/utils.ts]: Done

File: e2e/homepage.spec.ts (hunk 3 above)
Comment (matt-hernandez) [reply e2e/homepage.spec.ts]: Done

**Comment by matt-hernandez** [comment]
@brookslybrand I took a look at the file that is failing for the deploy. It has to do with allocating a dedicated IP in Fly for staging. I have two questions:

- Do we need to have a dedicated IP for staging? Do we need that extra power for an environment that isn't prod?
- I did a little research on the error and I think I have a solution. If you're okay with it, I can modify the workflow file to see if it works. Or would you rather leave that to @grantnorwood since he's the mastermind behind the CI/CD flow we have?

File: blog/writeFrontMatterCache.test.ts (hunk 1 above)
Comment (matt-hernandez) [reply blog/writeFrontMatterCache.test.ts:3]: I'm personally good with having tests before merging and getting good coverage. This is a learning experience for me and TBH I miss making stuff with others that feels high quality and airtight.

**Comment by brookslybrand** [comment]
> @brookslybrand The Github action said it had a problem deploying, but I can navigate to https://remixaustin-com-pr-99.fly.dev/ and see the site. So I'm guessing the error, while still needing to be fixed, didn't stop us from seeing the site with the blog. [And it appears to be working!!](https://remixaustin-com-pr-99.fly.dev/blog)

Weird, seems like some flag needs to be passed during this phase.

image

I can try to mess with it, but yeah, like you pointed out, it's working! Super cool 🙂 

Definitely an advantage of being on Fly, we can actually use the file system 😎. I do think eventually it might make more sense to leverage GitHub as our CMS like so Kent C Dodds and the Remix docs do, but for now seems like this will get us going.

File: blog/buildPosts.ts
Comment (matt-hernandez) [reply blog/buildPosts.ts:1]: I vote we save this for after the merge. Are you thinking of having a Node script or a bash script?

File: app/routes/resource/get-recent-front-matter.ts
Comment (matt-hernandez) [reply app/routes/resource/get-recent-front-matter.ts:1]: Done. This can be the start of a paginator control if/when we grow past 10 posts.

**Comment by brookslybrand** [comment]
> @brookslybrand I took a look at the file that is failing for the deploy. It has to do with allocating a dedicated IP in Fly for staging. I have two questions:
> 
> * Do we need to have a dedicated IP for staging? Do we need that extra power for an environment that isn't prod?
> * I did a little research on the error and I think I have a solution. If you're okay with it, I can modify the workflow file to see if it works. Or would you rather leave that to @grantnorwood since he's the mastermind behind the CI/CD flow we have?

Whoops, not sure how I missed this message originally

By all means push up your fix! I would prefer it to be a separate MR though, just to keep concerns separated. Would definitely like @grantnorwood to take a look at it either way.

File: app/routes/blog.$slug.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    <div className="container prose mx-auto px-4 py-8 md:px-0">
+      <h1 className="mb-[0px]">{title}</h1>
+      {author && (
+        <h4 className={`mt-4${date ? " mb-[0px]" : ""}`}>By {author}</h4>
+      )}
+      {date && (
+        <h5 className="mb-2">
+          {PUBLISH_DATE_FORMATTER.format(new Date(date))}
+        </h5>
+      )}
```
Comment (brookslybrand) [inline app/routes/blog.$slug.tsx:34]: Header levels should always be in order [without skipping](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#navigation)

```suggestion
        <h2 className={`mt-4${date ? " mb-[0px]" : ""}`}>By {author}</h2>
      )}
      {date && (
        <h3 className="mb-2">
          {PUBLISH_DATE_FORMATTER.format(new Date(date))}
        </h3>
      )}
```

File: app/routes/blog.$slug.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import { type LoaderArgs } from "@remix-run/node";
+import { useLoaderData } from "@remix-run/react";
+import { getMDXComponent } from "mdx-bundler/client";
+import { useMemo } from "react";
+import invariant from "tiny-invariant";
+import { PUBLISH_DATE_FORMATTER } from "~/utils";
+import { type Mdx } from "blog/parser";
```
Comment (brookslybrand) [inline app/routes/blog.$slug.tsx:7]: Nit: 

```suggestion
import  type { Mdx } from "blog/parser";
```

I didn't even realize this was possible 😅. It's really not a big deal, just adding the suggestion for consistency's sake.

File: app/routes/blog.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          ({ slug, title, date, author }, index: number) => (
+            <li key={`${title}-${index}`} className="pl-0">
+              <Link to={`/blog/${slug}`}>
+                <h3 className="mb-4 text-2xl font-bold">{title}</h3>
+                {(author || date) && (
+                  <h4>
+                    {author}
+                    {author && date && " - "}
+                    {date && PUBLISH_DATE_FORMATTER.format(new Date(date))}
+                  </h4>
```
Comment (brookslybrand) [inline app/routes/blog.tsx:29]: Same as above

```suggestion
                <h2 className="mb-4 text-2xl font-bold">{title}</h3>
                {(author || date) && (
                  <h3>
                    {author}
                    {author && date && " - "}
                    {date && PUBLISH_DATE_FORMATTER.format(new Date(date))}
                  </h3>
```

File: blog/paths.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import path from "path";
+
+export const POSTS_SOURCE_DIR = path.join(__dirname, ".", "posts");
+export const POSTS_BUILD_DIR = path.join(__dirname, "..", "public", "posts");
+export const FRONT_MATTER_CACHE_FILENAME = "front-matter-cache.json";
+export const PUBLIC_DIR = path.join(__dirname, "..", "public");
+export const FRONT_MATTER_CACHE_FILEPATH = path.join(
+  PUBLIC_DIR,
+  FRONT_MATTER_CACHE_FILENAME
+);
```
Comment (brookslybrand) [inline blog/paths.ts:10]: I'm gonna be honest, it feels a bit weird to be throwing mdx and tsx files into the public directory, just to later fetch them and build them on demand. Maybe this isn't a big deal, just trying to wrap my head around a partial build step that isn't really a build step.

Anyway, I'm not gonna stop this PR because of it, main because I would _way_ rather us have content than have the perfect blog engine. Long term I'd love to do something like the pattern of [fetching these files from GitHub](https://andre-landgraf.dev/blog/2022-05-29_how-to-integrate-markdown-content-with-syntax-highlighting-in-remix/#:~:text=Fetching%20Markdown%20files%20from%20GitHub) with some cache control headers to avoid the latency and cost of the bundler. However, I think this would require some more research and I should really write up an issue to get some feedback on this idea.

Anyway, tl;dr thanks again for doing this! I feel a bit weird about the approach, but you've got lots of great pieces here and I don't want to stop the momentum of getting this in. I will follow up with some issues when I have time to create them.

File: blog/writeFrontMatterCache.test.ts (hunk 1 above)
Comment (brookslybrand) [reply blog/writeFrontMatterCache.test.ts:3]: I might not be understanding right, but Colby are you saying your concerned that these 2 functions, which are used inside of `writeFrontMatterCache`, are exported just so they can be tested?

Personally I think this is fine if you want to test them, however I would say it's more valuable to just test `writeFrontMatterCache` and not worry about the implementation details. Your test probably shouldn't care about `validateFrontMatter` or any other implementation details of `writeFrontMatterCache`, just that `writeFrontMatterCache` is doing the right thing in all situations.

That's my opinion, but yeah, I don't think it's a big deal or a blocker. I do like the idea of following up with further testing holistically and improving coverage a bit.

File: app/utils.ts (hunk 2 above)
Comment (brookslybrand) [reply app/utils.ts]: Not asking you to revert this, but just a small thing: this isn't really a string constant, which is what we usually use SCREAM_CASE for. It's more akin to a function (or I guess object with methods). Anyway, not a big deal, I do think this is more consistent. Just personal preference is that SCREAM_CASE is used for string constants shared throughout the app.

**Review by brookslybrand — APPROVED** [review]

**Comment by brookslybrand** [comment]
Whoops, forgot to leave this comment.

@matt-hernandez I think this is good enough to merge, because I'm eager to get this in 🙂 

I have some follow up tickets I need to make to cleanup a few things and explore some more ideas.

However, even though I gave an approval, I'd really love for @colbywhite to have the final say on whether or not to merge this in, since he's been a bit more invested in the process.

Thanks y'all!

File: blog/paths.ts (hunk 7 above)
Comment (matt-hernandez) [reply blog/paths.ts:10]: Yeah, the original PR started out by building everything all the way. But what I discovered was that the final build of a blog post was quite large in file size. Granted, I used some images for my draft blog post, and even though they were optimized for size, they were base64 encoded and inserted into the blog data instead of being hosted somewhere. But even then, there is a lot of duplication for every blog post because mdx-bundler has to compile `tsx` into `js` and turn it into a component that can be consumed by the DOM. The amount of boilerplate in every blog post is hefty and I would argue the solution doesn't scale very well. By storing the source files for the blog in public and building them on demand, we reduce our footprint in `/public` by a lot and could, in theory, store a lot of blog posts without worrying about storage capacity.

I think all-in-all, we're seeing firsthand the benefits and drawbacks of a codebase as a CMS. The big benefit is that we can ship content super fast. The drawback is that we have to do some odd tricks to make it work and scale. If we decide to try and turn this POC into a package or stack for the whole Remix community, maybe we drop the idea of a file-based CMS and focus instead on using the Github API. Or maybe we give people the option to do both, but mention the drawbacks of each.

**Comment by matt-hernandez** [comment]
@colbywhite @brookslybrand I'm down if you all want to do a final huddle or code jam sesh before we merge. I can find time in my upcoming week.

**Comment by matt-hernandez** [comment]
@brookslybrand I met up with @colbywhite at a React meetup today. We're good with merging this. Tests will follow later.

**Comment by brookslybrand** [comment]
> @brookslybrand I met up with @colbywhite at a React meetup today. We're good with merging this. Tests will follow later.

Awesome! And bummed I couldn't make it last night, I had a conflict.

Pumped to see more tests and to review your first blog post 🤘

**Comment by matt-hernandez** [comment]
Well holy 💩 , I just won free tickets to Remix Conf from one of the conference sponsors. So it looks like I'll be going @brookslybrand lol. I'll get this post drafted ASAP. Want to make sure it's in good shape before the conference.

**Comment by brookslybrand** [comment]
> Well holy 💩 , I just won free tickets to Remix Conf from one of the conference sponsors. So it looks like I'll be going @brookslybrand lol. I'll get this post drafted ASAP. Want to make sure it's in good shape before the conference.

Wow that’s awesome! Looking forward to see you there and hanging out :)
